libyul/backends/wasm/EWasmCodeTransform.cpp

```diff
@@ -0,0 +1,317 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+* Common code generator for translating Yul / inline assembly to EWasm.
+*/
+
+#include <libyul/backends/wasm/EWasmCodeTransform.h>
+
+#include <libyul/backends/wasm/EWasmToText.h>
+#include <libyul/optimiser/NameCollector.h>
+
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Utilities.h>
+#include <libyul/Exceptions.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+string EWasmCodeTransform::run(Dialect const& _dialect, yul::Block const& _ast)
+{
+	EWasmCodeTransform transform(_dialect, _ast);
+	vector<wasm::FunctionDefinition> functions;
+
+	for (auto const& statement: _ast.statements)
+	{
+		yulAssert(
+			statement.type() == typeid(yul::FunctionDefinition),
+			"Expected only function definitions at the highest level."
+		);
+		functions.emplace_back(transform.translateFunction(boost::get<yul::FunctionDefinition>(statement)));
+	}
+
+	return EWasmToText{}.run(transform.m_globalVariables, functions);
+}
+
+wasm::Expression EWasmCodeTransform::generateMultiAssignment(
```

Comment:
Maybe simply always use global variables to return values? (Even when there is only one return value)

Not sure about the pros and cons.

Reply:
I would assume that global variables are way more expensive.